src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.cpp

```diff
@@ -313,362 +313,20 @@ INST_PhaseVolumeFractionKernel( 5, 3 );
-{
-  localIndex constexpr NDOF = NC + 1;
-
-  real64 const poreVol     = volume * porosity;
-  real64 const dPoreVol_dP = volume * dPoro_dPres;
-
-  localVolBalance = 1.0;
-  for( localIndex i = 0; i < NDOF; ++i )
-  {
-    localVolBalanceJacobian[i] = 0.0;
-  }
-
-  // sum contributions to component accumulation from each phase
-  for( localIndex ip = 0; ip < NP; ++ip )
-  {
-    localVolBalance -= phaseVolFrac[ip];
-    localVolBalanceJacobian[0] -= dPhaseVolFrac_dPres[ip];
-
-    for( localIndex jc = 0; jc < NC; ++jc )
-    {
-      localVolBalanceJacobian[jc+1] -= dPhaseVolFrac_dCompDens[ip][jc];
-    }
-  }
-
-  // scale saturation-based volume balance by pore volume (for better scaling w.r.t. other equations)
-  for( localIndex idof = 0; idof < NDOF; ++idof )
-  {
-    localVolBalanceJacobian[idof] *= poreVol;
-  }
-  localVolBalanceJacobian[0] += dPoreVol_dP * localVolBalance;
-  localVolBalance *= poreVol;
-}
-
-template< localIndex NC, localIndex NP >
-void
-VolumeBalanceKernel::
-  launch( localIndex const size,
-          globalIndex const rankOffset,
-          arrayView1d< globalIndex const > const & dofNumber,
-          arrayView1d< integer const > const & elemGhostRank,
-          arrayView1d< real64 const > const & volume,
-          arrayView2d< real64 const > const & porosity,
-          arrayView2d< real64 const > const & dPoro_dPres,
-          arrayView2d< real64 const, compflow::USD_PHASE > const & phaseVolFrac,
-          arrayView2d< real64 const, compflow::USD_PHASE > const & dPhaseVolFrac_dPres,
-          arrayView3d< real64 const, compflow::USD_PHASE_DC > const & dPhaseVolFrac_dCompDens,
-          CRSMatrixView< real64, globalIndex const > const & localMatrix,
-          arrayView1d< real64 > const & localRhs )
-{
-  forAll< parallelDevicePolicy<> >( size, [=] GEOSX_HOST_DEVICE ( localIndex const ei )
-  {
-    if( elemGhostRank[ei] >= 0 )
-      return;
-
-    localIndex constexpr NDOF = NC + 1;
-
-    real64 localVolBalance;
-    real64 localVolBalanceJacobian[NDOF];
-
-    compute< NC, NP >( volume[ei],
-                       porosity[ei][0],
-                       dPoro_dPres[ei][0],
-                       phaseVolFrac[ei],
-                       dPhaseVolFrac_dPres[ei],
-                       dPhaseVolFrac_dCompDens[ei],
-                       localVolBalance,
-                       localVolBalanceJacobian );
-
-    // get equation/dof indices
-    localIndex const localRow = dofNumber[ei] + NC - rankOffset;
-    globalIndex dofIndices[NDOF];
-    for( localIndex jdof = 0; jdof < NDOF; ++jdof )
-    {
-      dofIndices[jdof] = dofNumber[ei] + jdof;
-    }
+  class ElementBasedAssemblyKernel< NC, NDOF >
 
-    // TODO: apply equation/variable change transformation(s)
+INST_ElementBasedAssemblyKernel( 1, 2 );
+INST_ElementBasedAssemblyKernel( 2, 3 );
+INST_ElementBasedAssemblyKernel( 3, 4 );
+INST_ElementBasedAssemblyKernel( 4, 5 );
+INST_ElementBasedAssemblyKernel( 5, 6 );
```

Comment:
These explicit instantiations play no role, now that the entire class body is in the header and most assembly functions are additionally templated on the customization op type. Sadly, I don't see a "nice" way to retain the separate compilation benefits with this kernel design, but we should at least remove these to avoid the double instantiation cost.

Reply:
I removed these explicit instantiations. I will report here the compilation times from Travis where the checks are done. Hopefully the compilation times will remain ok with this new style.